rules/recording.go

```diff
@@ -89,7 +89,8 @@ func (rule *RecordingRule) Eval(ctx context.Context, ts time.Time, query QueryFu
 			lb.Set(l.Name, l.Value)
 		})
 
-		sample.Metric = lb.Labels(labels.EmptyLabels())
+		labelSetSize := len(sample.Metric) + len(rule.labels) + 1 // additional 1 for rule name
+		sample.Metric = lb.Labels(make(labels.Labels, 0, labelSetSize))
```

Comment:
We are working on abstracting labels away. You can not use make assuming it's a slice.

Reply:
As @roidelapluie says, this change is conflicting with other work to make `labels.Labels` much smaller, so should not go in as-is.

I imagine moving the `NewBuilder` out of the loop would give some improvement.

Why not just pass `sample.Metric` in to `lb.Labels()`?  That will work for the common case where the rule doesn't add its own labels.

Reply:
> I imagine moving the `NewBuilder` out of the loop would give some improvement.
> 
> Why not just pass `sample.Metric` in to `lb.Labels()`? That will work for the common case where the rule doesn't add its own labels.

Good ideas - these both improve things even further:

```
name                                                                                old time/op    new time/op    delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  8.62µs ± 0%    8.16µs ± 0%  -5.26%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            8.76µs ± 0%    8.45µs ± 0%  -3.54%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    8.56µs ± 0%    8.18µs ± 0%  -4.53%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               16.3µs ± 0%    15.9µs ± 0%  -2.08%  (p=0.008 n=5+5)

name                                                                                old alloc/op   new alloc/op   delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  7.37kB ± 0%    6.68kB ± 0%  -9.35%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            7.37kB ± 0%    7.07kB ± 0%  -4.12%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    7.18kB ± 0%    6.68kB ± 0%  -6.91%  (p=0.000 n=5+4)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               15.5kB ± 0%    15.1kB ± 0%  -2.79%  (p=0.008 n=5+5)

name                                                                                old allocs/op  new allocs/op  delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                     134 ± 0%       126 ± 0%  -5.97%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10               134 ± 0%       128 ± 0%  -4.48%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10       132 ± 0%       126 ± 0%  -4.55%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10                  246 ± 0%       240 ± 0%  -2.44%  (p=0.008 n=5+5)
```